File: java/src/test/java/kms/history/TestGetPotentialHistory.java
package kms.history;

import dev.spiralmoon.maplestory.api.common.MapleStoryApiErrorCode;
import dev.spiralmoon.maplestory.api.common.MapleStoryApiException;
import dev.spiralmoon.maplestory.api.kms.MapleStoryApi;
import dev.spiralmoon.maplestory.api.kms.dto.history.PotentialHistoryResponseDTO;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.*;

import java.time.LocalDateTime;

@DisplayName("getPotentialHistory")
public class TestGetPotentialHistory {
    private static final String apiKey = System.getProperty("API_KEY_KMS"); // Your API Key
    private static final MapleStoryApi api = new MapleStoryApi(apiKey);

    @Test
    @DisplayName("success: getPotentialHistory")
    void getPotentialHistory() {
        int count = 1000;
        PotentialHistoryResponseDTO response = api.getPotentialHistory(count).join();
        System.out.println(response.toString());
    }

    @Test
    @DisplayName("success: getPotentialHistory with date")
    void getPotentialHistory_with_date() {
        int count = 1000;
        LocalDateTime date = LocalDateTime.of(2025, 7, 31, 0, 0);
        PotentialHistoryResponseDTO response = api.getPotentialHistory(count, date).join();
        System.out.println(response.toString());
    }

    @Test
    @DisplayName("success: getPotentialHistory with cursor")
    void getPotentialHistory_with_cursor() {
        int count = 10;
        LocalDateTime date = LocalDateTime.of(2025, 7, 31, 0, 0);
        PotentialHistoryResponseDTO response = api.getPotentialHistory(count, date).join();
        String nextCursor = response.getNextCursor();
        assertThat(nextCursor).isNotNull();

        PotentialHistoryResponseDTO nextResponse = api.getPotentialHistory(count, nextCursor).join();
        System.out.println(nextResponse.toString());
    }

    @Test
    @DisplayName("fail: getPotentialHistory with invalid date")
    void getPotentialHistory_with_invalid_date() {
        int count = 1000;
        LocalDateTime invalidDate = LocalDateTime.of(2024, 1, 24, 0, 0);
        assertThatThrownBy(() -> api.getPotentialHistory(count, invalidDate).join())
                .hasCauseInstanceOf(IllegalArgumentException.class)
                .satisfies(e -> {
                    Throwable error = e.getCause();
                    assertThat(error.getMessage()).contains("You can only retrieve data after 2024-01-25.");
                    System.out.println(error.getMessage());
                });
    }

    @Test
    @DisplayName("fail: getPotentialHistory with invalid cursor throw OPENAPI00003")
    void getPotentialHistory_with_invalid_cursor() {
        int count = 1000;
        String invalidCursor = "invalid_cursor_123";
        assertThatThrownBy(() -> api.getPotentialHistory(count, invalidCursor).join())
                .hasCauseInstanceOf(MapleStoryApiException.class)
                .satisfies(e -> {
                    MapleStoryApiException apiException = (MapleStoryApiException) e.getCause();
                    assertThat(apiException.getErrorCode()).isEqualTo(MapleStoryApiErrorCode.OPENAPI00003);
                    System.out.println(apiException.getErrorCode() + " " + apiException.getMessage());
                });
    }

    @Test
    @DisplayName("fail: getPotentialHistory with excessive count throw OPENAPI00004")
    void getPotentialHistory_with_excessive_count() {
        int count = 9999;
        assertThatThrownBy(() -> api.getPotentialHistory(count).join())
                .hasCauseInstanceOf(MapleStoryApiException.class)
                .satisfies(e -> {
                    MapleStoryApiException apiException = (MapleStoryApiException) e.getCause();
                    assertThat(apiException.getErrorCode()).isEqualTo(MapleStoryApiErrorCode.OPENAPI00004);
                    System.out.println(apiException.getErrorCode() + " " + apiException.getMessage());
                });
    }
}

File: csharp/MapleStory.Test/KMS/History/TestGetPotentialHistory.cs
using NUnit.Framework;
using MapleStory.OpenAPI.KMS;
using MapleStory.OpenAPI.Common;

namespace MapleStory.Test.KMS.History
{
    [TestFixture]
    public class TestGetPotentialHistory
    {
        private static readonly string apiKey = EnvConfig.Get("API_KEY_KMS"); // Your API Key
        private static readonly MapleStoryAPI api = new MapleStoryAPI(apiKey);

        [Test, Description("success: GetPotentialHistory")]
        public async Task GetPotentialHistory()
        {
            int count = 1000;
            var response = await api.GetPotentialHistory(count);
            Console.WriteLine(response.ToJson());
        }

        [Test, Description("success: GetPotentialHistory with date")]
        public async Task GetPotentialHistory_With_Date()
        {
            int count = 1000;
            var date = new DateTime(2025, 7, 31, 0, 0, 0);
            var response = await api.GetPotentialHistory(count, date);
            Console.WriteLine(response.ToJson());
        }

        [Test, Description("success: GetPotentialHistory [with cursor]")]
        public async Task GetPotentialHistory_With_Cursor()
        {
            int count = 10;
            var date = new DateTime(2025, 7, 31, 0, 0, 0);
            var response = await api.GetPotentialHistory(count, date);
            var nextCursor = response.NextCursor;
            Assert.That(nextCursor, Is.Not.Null);
            var nextResponse = await api.GetPotentialHistory(count, nextCursor);
            Console.WriteLine(nextResponse.ToJson());
        }

        [Test, Description("fail: GetPotentialHistory with invalid date")]
        public void GetPotentialHistory_With_Invalid_Date()
        {
            int count = 1000;
            var invalidDate = new DateTime(2024, 1, 24, 0, 0, 0);
            var e = Assert.ThrowsAsync<ArgumentException>(async () => await api.GetPotentialHistory(count, invalidDate));
            Assert.That(e.Message, Does.Contain("You can only retrieve data after 2024-01-25."));
            Console.WriteLine(e.Message);
        }

        [Test, Description("fail: GetPotentialHistory [throws OPENAPI00003 with invalid cursor]")]
        public void GetPotentialHistory_With_Invalid_Cursor_Throws_OPENAPI00003()
        {
            int count = 1000;
            var invalidCursor = "invalid_cursor_123";
            var e = Assert.ThrowsAsync<MapleStoryAPIException>(async () => await api.GetPotentialHistory(count, invalidCursor));
            Assert.That(e.ErrorCode, Is.EqualTo(MapleStoryAPIErrorCode.OPENAPI00003));
            Console.WriteLine($"{e.ErrorCode} {e.Message}");
        }

        [Test, Description("fail: GetPotentialHistory [throws OPENAPI00004 with excessive count]")]
        public void GetPotentialHistory_With_Excessive_Count_Throws_OPENAPI00004()
        {
            int count = 9999;
            var e = Assert.ThrowsAsync<MapleStoryAPIException>(async () => await api.GetPotentialHistory(count));
            Assert.That(e.ErrorCode, Is.EqualTo(MapleStoryAPIErrorCode.OPENAPI00004));
            Console.WriteLine($"{e.ErrorCode} {e.Message}");
        }
    }
}
